### DISCUSSION TITLE
Exposing plugin configurations

### DISCUSSION THREAD

**Discussion by zmzlois** [rolldown/rolldown#2577] [Ideas]
### Current behavior 
When we are processing user's configuration in Vite (tried on both rollup and rolldown), there isn't any ways for us to process user's configuration from other plugin ("we", "us" as a plugin)

#### Example with vite(rollup and rolldown) running build with `svgr()` options: 
![Screenshot 2024-11-03 at 19 58 15](https://github.com/user-attachments/assets/95bdcddb-eb1c-4db9-b1c7-24e9c06ecb6f)
![Screenshot 2024-11-03 at 19 57 27](https://github.com/user-attachments/assets/7053c2ad-15bb-4cde-8e83-d2063ed0dd8f)

#### Exmple with ModuleFederationPlugin in webpack with module federation options: 
![Screenshot 2024-11-03 at 20 29 46](https://github.com/user-attachments/assets/ee3f57d9-36d4-4ca8-82d6-32e8521efbad)
![Screenshot 2024-11-03 at 20 36 25](https://github.com/user-attachments/assets/5ff8b126-fd0b-4bbd-847b-46c7cb3585eb)

[Repo for this comparison](https://github.com/zmzlois/bundler-compare)
### Proposed behavior 
If user passed in configuration options to other plugins, it should be available for other plugins, or somewhere can be tapped into to see, keeping `load: [Function: load]` is cool atm. 

### Why is this beneficial 
1. For large number of enterprises, their internal requirement to analyse build statistics, what plugins are being used for securities reasons, build performance tracking, analytics are high - it will help greatly when there is such needs 
2. Benefits toolings that are built around plugin ecosystems in build tools

### Reference in Webpack 
Q1. How does webpack fundamentally prevent ordering issues? 
No sequencing of "pre" or "post", the sequence relying on the lifecycle hooks. Plugin author could keep a statistics track of bundling process, lifecycle hook processing, even it is the same lifecycle hook, you could be handle it similar to a `z-index` on how sooner or later this hook should run. 

[Reference](https://webpack.js.org/api/plugins/#reporting-progress)

**Comment by RussellCanfield** [comment]
I believe Vite has the config and configResolved methods that could be used here as a potential solution here. Technically you would build your config in the config function, and then you publish it to Vite's global and its available on configResolved.

**Reply by zmzlois** [reply]
but if it is handled this way we will need to PR federation to expose configuration options? 

we need to tackle two problems: 
1.  sequence of "post" and "pre" from two plugins under one - any other plugin has "pre" that interfere build output with `load` or `transform` breaking federation,
2. making sure on `closeBundle` is actually at the end of all plugins to ensure all other plugins are correctly ran and we are not missing any required transformation 

ideally if configuration options can be accessed from `configResolved` 
```ts
plugin: [
   federation({ 
      //  federation option 
  }), 
   svgr(), // let other `pre` and `transform` run
      ... // rest of the plugins, 
   { 
     name: 'other-plugin' 
     enforce: "post", 
     transform: (id, code) => {} // handle federation option 
     closeBundle: () => {} // handle when all asset sealed, good to even have a statistic track of progress: "1000" to make sure this is the last `closeBundle` across plugins
   }
```

**Comment by yyx990803** [comment]
Just to clarify, there seem to be two separate issues that are not fundamentally related here:

1. Plugins being able to access other plugins' config options.
2. Sequencing of plugins, and specifically what Zephyr's plugin is trying to do.

I think we should address these separately - also, I assume we are discussing this specifically in the context of the Zephyr plugin which wraps the Vite MF plugin?

## Plugins being able to access other plugins' config options

The fundamental issue here is that Rollup / Vite / Rolldown plugins are just objects - and exposing a factory function that take options is just a convention. Which means most plugins just return the actual plugin object without storing the options anywhere.

Assuming the use case here is for the Zephyr plugin to read MF plugin's config, then it's really up to the MF plugin to expose the options somewhere - it can be just an extra property like `_config` attached to one of its sub plugins, e.g. the one with name `module-federation-vite`. And then the Zephyr plugin can search for it.

## Plugin Sequencing

Again I assume what you want is the following order:

```
federation
...other user plugins
zephyr's post plugin
```

You want `federation` to run before all other user plugins, while some Zephyr plugins to run after all other user plugins. If that is the case, one thing you can do is to change `withZephyr` to receive and augment the entire plugin array instead:

```js
function withZephyr(options, plugins) {
  return [
    federation(),
    ...plugins,
    zephyr()
  ]
}
```

Then in config:

```js
export default {
  plugins: withZephyr({ /* ... */ }, [
    svgr()
  ])
}
```

**Reply by gioboa** [reply]
Thanks for your exhaustive explanation @yyx990803

**Comment by zmzlois** [comment]
> I think we should address these separately - also, I assume we are discussing this specifically in the context of the Zephyr plugin which wraps the Vite MF plugin?

Agree and yes :) 

> it can be just an extra property like _config attached to one of its sub plugins, e.g. the one with name module-federation-vite. And then the Zephyr plugin can search for it.

Thanks for the answer - fixed waiting to get merged (the type error scared everyone to do it 😭): https://github.com/module-federation/vite/pull/175

> You want federation to run before all other user plugins, while some Zephyr plugins to run after all other user plugins. If that is the case, one thing you can do is to change withZephyr to receive and augment the entire plugin array instead
```ts
function withZephyr(options, plugins) {
  return [
    federation(),
    ...plugins,
    zephyr()
  ]
}
```

 second answer is super helpful as well 🚀 🙏 will keep you posted this is great!!

**Reply by zmzlois** [reply]
to add notes on why mention sequence: 

it was because we weren't able to take user options and needing to pin federation under zephyr to take user config: 

```ts
function withZephyr(_options) {
    return federation(_options).concat([zephyr(_options)])
           --------------------      ------------------
                 "pre"                         "post" 
```

and it resulting in a mess of plugin sequence for users: 

```ts
plugins: [
 withZephyr(config), // "pre" and "post" 
 svgr(), // ...other plugins with "pre"
```

Once we are able to parse the configuration from users using federation we can separate zephyr plugin and federation back to where they should be
